design/batch_retrieval_design.md

```diff
@@ -2,19 +2,140 @@
+| 2    | *Server side:*<br />Update Batch retrieval to return list of variables **and** their response (success/failure) (**) | - Will help us during rotation for Milestone 2<br />- Best practice / straightforward design for batch endpoint | - Requires both client/server implementation<br /> | M, 10 days            |
+| 3    | Stay as is                                                   | No additional work needed                                    | Bad UX                                             | -                     |
 
 (*) Fallback solution: use this solution as a safety net, only if the original Batch Retrieval request fails
 
 (**) This solution can be broken up in two: 1. Create a new API endpoint or 2. use existing one (return 409 response)
 
+*Decision:* We decided to go with #2 and create our own API. The new batch retrieval API endpoint will not be forward facing and be for **internal use only**. 
+
+### Design
+
+#### Secrets Provider Job flow
+
+As a small recap, the following explains the Secrets Provider Job flow
+
+1. The `cyberark-secrets-provider-for-k8s` runs as a Job and authenticates with the Conjur/DAP Follower using the Kubernetes Authenticator (`authn-k8s`).
+
+2. The `cyberark-secrets-provider-for-k8s` reads all Kubernetes secrets required by the applications in the same namespace.
+
+3. For each mapped Kubernetes secret, 
+
+   a. The `cyberark-secrets-provider-for-k8s` performs a Batch Retrieval request and retrieves Conjur/DAP secrets.
+
+   b. It updates the Kubernetes secret with the Conjur/DAP secret value.
+
+4. The `cyberark-secrets-provider-for-k8s` Job is completed.
+
+The new Batch Retrieval API will effect #2a. 
+
+### URI
+
+##### POST /secrets/batch
+
+This method retuns a list of secrets, their status code, and their value if an error did not occur. 
+
+```
+POST /secrets/batch
+
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+### Request Body
+
+The request body is a list of all variable IDs (paths to secrets in Conjur/DAP) that need to be retrieved. For example: 
+
+```
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+#### Headers
+
+```
+Content-Type: application/json
+```
+
+### Response
+
+| Code             | Description                                                  |
+| ---------------- | ------------------------------------------------------------ |
+| 200 OK           | All secret values were retrieved successfully                |
+| 400 Bad Request  | Server cannot process request due to client error (invalid input parameters or logical validations) |
+| 401 Unauthorized | The request lacks valid authentication credentials           |
+| 403 Forbidden    | The authenticated user lacks the necessary privilege         |
```

Comment:
for all the secrets? what happens if the user has permissions for one secret but not for another?

Reply:
I think I explained this below but all secret values will be treated as individual entities. If a single secret fails to be retrieved, this will not impact the integrity of the request. I will add this statement in this Response section.

Reply:
so when will we return 403? only if all the secrets are forbidden? can you emphasize this in the table above?

Reply:
also - you return 404 if at least one is not found but 403 if all the secrets are not found. i think this is lacking consistency.

Reply:
No no if we can parse the body of the request, the response will be 200 even if the sub-requests have failed with 401 / 404

Reply:
then i'll ask again. when will we return 403 - when all the secrets are forbidden? when at least one of the secrets is forbidden?

can you state this in the table?

Reply:
They are already separated like so:
### Response
## Batch Response
 # error codes (200 400)

## Individual secret responses
# error codes (200, 400, 401, 403, 404)

Was that not clear enough?

Reply:
@sigalsax I meant to create a different table for each.
Since Batch Responses are the API response code, and the Individual responses are in the response body, they are inherently different.
So they should be described differently.
The Batch Responses should be described in the API response code section, but the Individual responses should be described in the Response Body section. and a table for each for their relevant values and meaning.